### PR TITLE
boards: opta: initial support for M7 core

### PR DESCRIPTION
**Pull request by fogzot**
Features enabled by this commit:
* M7 core available with or without the Arduino MCU-based bootloader by correctly initializing the external clock source.
* All LEDs, relays and the user button are mapped into the device tree (samples/common/blinky works out of the box).
* USB can be used in ACM mode (samples/subsys/usb/cdc_acm works without changes).

### CODE REVIEW

File: boards/arduino/opta/arduino_opta-common.dtsi
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta-common.dtsi:38]: this file's formatting needs sorting

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4.dts
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m4.dts:3]:
```suggestion
 * Copyright (c) 2024 DNDG srl
 *
```

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:32]: console is required

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:35]: ?

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:5]:
```suggestion
# Enable GPIO
```

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:8]:
```suggestion
# Clock configuration
```

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:17]:
```suggestion
# Enable pin control
```

File: boards/arduino/opta/board_gpio_hse.c
Comment (nordicjm) [inline boards/arduino/opta/board_gpio_hse.c:23]: comment as to what this does might be useful?

File: boards/arduino/opta/board_gpio_hse.c
Comment (nordicjm) [inline boards/arduino/opta/board_gpio_hse.c:19]: comment can go?

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: boards/arduino/opta/arduino_opta-common.dtsi
Comment (fogzot) [reply boards/arduino/opta/arduino_opta-common.dtsi:38]: What do you mean by "sorting"? I see that I used spaces in a file that originally used tabs: fixed that but I don't if it was what you wanted.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig
Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:32]: The Opta has a single USB that can be used as console if configured as ACM. It can also work as USB storage (I did not yet try this on Zephyr): does it make sense to always enable the console even if the application can configure the USB in a different way?

Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:32]: I just pushed a new version that enables the console, but please see my question above.

File: boards/arduino/opta/board_gpio_hse.c
Comment (nordicjm) [inline boards/arduino/opta/board_gpio_hse.c:17]: `/* */` style comments, tab indent

File: boards/arduino/opta/doc/index.rst
Comment (nordicjm) [inline boards/arduino/opta/doc/index.rst:67]: board target, use proper board name with \`\` around it i.e. `arduino_opta/stm32h747xx/m4`

File: boards/arduino/opta/doc/index.rst
Comment (nordicjm) [inline boards/arduino/opta/doc/index.rst:48]: as other comment, board target, use \`\` and `arduino_opta/stm32h747xx/m7`

**Review by nordicjm — COMMENTED** [review]
Some minor nit

File: boards/arduino/opta/arduino_opta-common.dtsi
Comment (nordicjm) [reply boards/arduino/opta/arduino_opta-common.dtsi]: That was it

**Review by pillo79 — COMMENTED** [review]
Thanks for this PR! LGTM, just minor nits, will try next week on an actual Opta :+1:

File: boards/arduino/opta/doc/index.rst
Comment (nordicjm) [inline boards/arduino/opta/doc/index.rst:48]:
```suggestion
The ``arduino_opta/stm32h747xx/m7`` board target
```

File: boards/arduino/opta/doc/index.rst
Comment (nordicjm) [inline boards/arduino/opta/doc/index.rst:69]:
```suggestion
The ``arduino_opta/stm32h747xx/m4`` board target
```

**Comment by fogzot** [comment]
@pillo79 @nordicjm there is still something I should do to get this merged? I have some other work already done (Ethernet and RS485) but I don't want to restart the whole process and I'd like to have this one merged before moving to the next steps.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: this must not be done at board level, drivers must do this

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig
Comment (nordicjm) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:18]: same here

**Review by nordicjm — DISMISSED** [review]

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: Sorry, what exactly should not be done? If you mean enabling pin control, I just copied this from other STM32 boards, see for example: `zephyr/boards/st/stm32h747i_disco/stm32h747i_disco_stm32h747xx_m7_defconfig`. The same is true for `CONFIG_HW_STACK_PROTECTION`.

Comment (nordicjm) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: pinctrl

Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: But if I remove it even the simplest sample (blinky) fails because of missing symbols:
```
/home/fog/Workspaces/zephyrproject/zephyr/include/zephyr/drivers/pinctrl.h:350: undefined reference to `pinctrl_lookup_state'
```
I checked 10 random boards in `zephyr/boards/st` and all have `CONFIG_PINCTRL=y` in defconfig. Given that the Opta is basically an stm32h747i why can't it be configured the same way?

Comment (nordicjm) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: Because that is wrong, @gmarull can explain

Comment (nordicjm) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: Also @erwango you need to fix your drivers

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4.dts
Comment (pillo79) [inline boards/arduino/opta/arduino_opta_stm32h747xx_m4.dts:33]: 512K is the correct value for the M4 flash size. Why is this change required?

**Review by pillo79 — COMMENTED** [review]
Tested this PR successfully on the Opta today! :partying_face: Several small nits, but the Flash sizes and `CONFIG_SERIAL` are important points to discuss.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (pillo79) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: @fogzot To add some context, IIUC, the proper solution would be to have this symbol automatically `select`ed via Kconfig by drivers that need it, like this USB driver for some NXP device does:  

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/usb/device/Kconfig#L166-L173

The STM32 USB driver does not seem to do that:

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/usb/device/Kconfig#L46-L56

... even though it uses it internally, e.g.:

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/usb/device/usb_dc_stm32.c#L444

This is an issue with some current Zephyr drivers that should be fixed treewide. Personally I don't see this as a reason to block a PR adding support for additional cores, but that's just my opinion.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7.dts
Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m7.dts:81]: I suppose you want to use the 75%-25% split of the 2M flash between M7 and M4 as the Arduino IDE does? This is one of the things I was unsure about how to proceed (how do you declare a size of 1280 given that the slot0 size is only 1024? STM device tree entries list the flash as 2 slots of 1024 each...) so I just erred on the safe side and i didn't "spill" into the second flash bank.
If it is OK to declare a size larger than the slot I see no problems with the change you want.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig
Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig:29]: You are right. I removed it in a past commit by mistake.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4.dts
Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4.dts:33]: My mistake.

File: boards/arduino/opta/doc/index.rst
Comment (fogzot) [reply boards/arduino/opta/doc/index.rst:156]: Is this OK?
```
By default:

  - CPU1 (Cortex-M7) boot address is set to 0x08040000
  - CPU2 (Cortex-M4) boot address is set to 0x08180000

Zephyr flash configuration has been set to be compatible with the 
"Flash split: 1.5MB M7 + 0.5MB M4" option in the Arduino IDE. The flash is 
partitioned as follows:

  - 0x08000000-0x0803FFFF Arduino MCUboot-derived bootloader
  - 0x08040000-0x0817FFFF M7 application
  - 0x08040000-0x0817FFFF M4 application
```

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: I understand the problem but if I apply the requested change my pull request will never pass the automatic checks because the tests won't build. Sincerely I don't like the idea of putting this on hold until the drivers are fixed but I don't see how to proceed otherwise. Suggestions very welcome.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7.dts
Comment (pillo79) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m7.dts:81]: My mistake, it is not possible to define a partition spanning multiple Flash devices and they have been assigned to minimize contention. So `768` is OK here!

File: boards/arduino/opta/arduino_opta_stm32h747xx_m7.yaml
Comment (pillo79) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m7.yaml:10]: Disregard this, see comment above

File: boards/arduino/opta/doc/index.rst
Comment (pillo79) [reply boards/arduino/opta/doc/index.rst:156]: Considering the above discussion, I'd say:
```
  - 0x08000000-0x0803FFFF (256k) Bootloader
  - 0x08040000-0x080FFFFF (768k) M7 application
  - 0x08180000-0x081FFFFF (512k) M4 application
```

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (erwango) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: > Also @erwango you need to fix your drivers

@nordicjm 

On the form:
This feels as an aggressive statement. Please rephrase.

On the content:
Ok, sure this can be done. And I agree it makes sense, but it isn't a but per se and I think you'll agree this is not linked to this board, so no reason to block.
Note also that STM32 is not an exception here: TI, ADI, Xilinx, Infineon, Microchip, Nuovoton do the same and some use of this could aslo be found on Nordic boards:
https://github.com/zephyrproject-rtos/zephyr/blob/a2f581f99459b3c9673f4403c346f9100952554a/boards/nordic/nrf7002dk/nrf7002dk_nrf5340_cpunet_defconfig#L13

**Comment by uLipe** [comment]
@fogzot I was not aware of your PR before and submit the same thing, it seems you made a great progress here so I will close mine, feel free to pick something from there that may be used to enhance your PR.

**Comment by fogzot** [comment]
@uLipe sure, lets not duplicate efforts. I also have ETH working but I have not yet published that branch (it depends on this pull request). If you want I can publish it so that we can cooperate. Working on RS485 right now.

**Comment by uLipe** [comment]
Sounds like a good plan, I'll be closing mine, and let's go togheter on the improving commits.

I have been working also on the wifi addition, but it depends on some work from ST guys, I will share the branch as well since it is the same solution for Giga, Nicla and Opta.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (erwango) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: @nordicjm Feedback on my comment ?

**Comment by fogzot** [comment]
I rebased and pushed a new version that should address all open requests *except* for the removal of `CONFIG_PINCTRL` (because some checks will fail otherwise). I'll let you decide if to merge or not.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (gmarull) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: Please, open an issue to fix STM32 platform. All drivers using pinctrl should select the option. I guess it's ok to merge provided this gets fixed soon.

File: boards/arduino/opta/doc/index.rst
Comment (uLipe) [inline boards/arduino/opta/doc/index.rst:116]: @fogzot would be possible to add here the mapping of the inputs as I did in my PR?

**Review by uLipe — CHANGES_REQUESTED** [review]

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (erwango) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: > Please, open an issue to fix STM32 platform. All drivers using pinctrl should select the option. I guess it's ok to merge provided this gets fixed soon.

@gmarull Sure, working on it.
@nordicjm ^^ An example of an appropriate manner to request such change

**Comment by fogzot** [comment]
@uLipe I have added the input mapping to my WIP branch and I'll surely grab other things from your changes as I go on. Right now I have ethernet working and I plan to have RS485 soon. You can find the branch here:
https://github.com/dndg/zephyr/tree/opta_wip

I think it would be better to coordinate outside of the PR, to avoid adding too much noise. You can find me on Discord (fogzot) or by email.

**Review by pillo79 — DISMISSED** [review]
LGTM!

**Review by uLipe — DISMISSED** [review]

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (uLipe) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig:16]: @erwango , @nordicjm , @gmarull  can we resolve this? It seems the only thing missing to move forward on merging this PR.

**Comment by uLipe** [comment]
Tested the PR on my Opta unit, and its working flawless.

**Comment by uLipe** [comment]
@fogzot , could you please rebase to latest main? Thank you!

**Comment by fogzot** [comment]
> @fogzot , could you please rebase to latest main? Thank you!

Done. I also removed the `CONFIG_PINCTRL` as was done for M4.

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (erwango) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig]: Solved.

**Review by erwango — APPROVED** [review]

File: boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig
Comment (fogzot) [reply boards/arduino/opta/arduino_opta_stm32h747xx_m4_defconfig]: Removed `CONFIG_PINCTRL`.

**Review by nordicjm — APPROVED** [review]

File: boards/arduino/opta/doc/index.rst
Comment (pillo79) [inline boards/arduino/opta/doc/index.rst:167]: Addresses in the newly added doc are still wrong.

**Review by pillo79 — COMMENTED** [review]
@fogzot Note there are still unfixed notes with the documentation. Given the troubled past of this PR, I'm not gonna be the one to stop the merge :hand_over_mouth:  - but please open a separate PR with the doc fix as soon as this is merged.

In the future, please do not mark conversations as "resolved" yourself, leave that to reviewers unless they are trivial (that was not!). Thanks! :slightly_smiling_face:

**Review by uLipe — APPROVED** [review]

**Review by pillo79 — APPROVED** [review]

**Pull request merged**
